bricks/_common/mphalport.c

```diff
@@ -71,14 +73,29 @@ mp_uint_t mp_hal_stdout_tx_strn(const char *str, size_t len) {
             return len - remaining;
         }
 
-        MICROPY_EVENT_POLL_HOOK
+        // Allow long prints to be interrupted.
+        if (remaining) {
+            MICROPY_EVENT_POLL_HOOK
+        }
     }
 
     return len;
 }
 
 void mp_hal_stdout_tx_flush(void) {
+    // Don't raise, just wait for data to clear.
     while (!pbsys_host_tx_is_idle()) {
-        MICROPY_EVENT_POLL_HOOK
+        MICROPY_VM_HOOK_LOOP;
+    }
+
+    // A program may be interrupted in the middle of a long print, or the user
+    // may have printed without a newline. Ensure we end on a new line.
+    if (!ended_on_new_line) {
+        const char *eol = "\r\n";
+        uint32_t size = strlen(eol);
+        pbsys_host_stdout_write((const uint8_t *)eol, &size);
```

Comment:
What if this fails or only does a partial write? If there is a good reason to ignore it, we should have a comment that explains why.